File: QtClient/QtClimaClient/Network/GenericServices/ProductionService.cpp
#include "ProductionService.h"

#include <Network/GenericServices/Messages/PreparingConfigRequest.h>
#include <Network/GenericServices/Messages/ProductionConfigRequest.h>
#include <Network/GenericServices/Messages/ProductionStateResponse.h>

ProductionService::ProductionService(QObject *parent) : INetworkService(parent)
{

}

void ProductionService::GetProductionState()
{
    NetworkRequest *request = new NetworkRequest();
    request->jsonrpc = "0.1a";
    request->service = "ProductionService";
    request->method = "GetProductionState";

    emit SendRequest(request);
}

void ProductionService::StartPreparing(float temperature, QDateTime startDate)
{
    NetworkRequest *request = new NetworkRequest();
    request->jsonrpc = "0.1a";
    request->service = "ProductionService";
    request->method = "StartPreparing";
    PreparingConfigRequest cfg;
    cfg.Config.TemperatureSetPoint = temperature;
    cfg.Config.StartDate = startDate;
    request->params = cfg.toJsonString();
    emit SendRequest(request);
}

void ProductionService::StartProduction(int placeHeads, QDateTime plandingDate, QDateTime startDate)
{
    NetworkRequest *request = new NetworkRequest();
    request->jsonrpc = "0.1a";
    request->service = "ProductionService";
    request->method = "StartProduction";

    ProductionConfigRequest cfg;
    cfg.Config.PlaceHeads = placeHeads;
    cfg.Config.PlandingDate = plandingDate;
    cfg.Config.StartDate = startDate;

    request->params = cfg.toJsonString();

    emit SendRequest(request);
}

void ProductionService::StopProduction()
{
    NetworkRequest *request = new NetworkRequest();
    request->jsonrpc = "0.1a";
    request->service = "ProductionService";
    request->method = "StopProduction";

    emit SendRequest(request);
}

QList<QString> ProductionService::Methods()
{
 return QList<QString>();
}

void ProductionService::ProcessReply(const NetworkResponse &reply)
{
    if(reply.service == "ProductionService")
    {
        ProductionStateResponse resp;
        resp.fromJson(reply.result.toUtf8());

        if(reply.method == "StartPreparing")
        {
            emit PreparingStarted(resp.State.State);
        }
        else if(reply.method == "StartProduction")
        {
            emit ProductionStarted(resp.State.State);
        }
        else if(reply.method == "StopProduction")
        {
            emit ProductionStopped(resp.State.State);
        }

        emit ProductionStateChanged(resp.State);
    }
}

File: ClimaDaemon/Core/Clima.Core.Scheduler/Network/Services/ProductionService.cs
using Clima.Basics.Services.Communication;
using Clima.Core.DataModel;
using Clima.Core.Network.Messages;
using Clima.Core.Scheduler.Network.Messages;
using Clima.Core.Scheduler;

namespace Clima.Core.Scheduler.Network.Services
{
    public class ProductionService:INetworkService
    {
        private readonly IClimaScheduler _scheduler;

        public ProductionService(IClimaScheduler scheduler)
        {
            _scheduler = scheduler;
        }

        [ServiceMethod]
        public ProductionStateResponse GetProductionState(DefaultRequest request)
        {
            return CreateResponse();
        }

        [ServiceMethod]
        public ProductionStateResponse StartPreparing(PreparingConfigRequest request)
        {
            _scheduler.StartPreparing(request.Config);
            return CreateResponse();
        }

        [ServiceMethod]
        public ProductionStateResponse StopProduction(DefaultRequest request)
        {
            _scheduler.StopProduction();
            return CreateResponse();
        }

        [ServiceMethod]
        public ProductionStateResponse StartProduction(DefaultRequest request)
        {
            _scheduler.StartProduction();
            return CreateResponse();
        }
        private ProductionStateResponse CreateResponse()
        {
            
            return new ProductionStateResponse()
            {
                State = new ProductionState()
                {
                    State = (int) _scheduler.SchedulerState,
                    StartDate = _scheduler.StartDate,
                    CurrentDay = _scheduler.CurrentDay,
                    CurrentHeads = _scheduler.CurrentHeads
                }
            };
        }
    }
}